// loy/build/module/index.js
import { createTestObject, getItem } from './util';
export const Test = '<>!-!_LOI_GENERATE_DESCRIPTIONS_!-!<>';
export const DateItem = 'loi_internal_Date';
export const Url = 'loi_internal_Url';
export const File = 'loi_internal_File';
export const Image = 'loi_internal_Image';
export const Enum = 'loi_internal_Enum';
/**
 * Return a result
 * @param result expected result
 * @param user user object
 * @param state state
 */
export function Give(result, user, state = { permissions: [], examples: [], type: [] }) {
    return {
        // returns value
        ok() {
            if (result === Test) {
                const { previous, ...rest } = state;
                return {
                    ...getItem(rest.permissions, "permissions"),
                    ...getItem(rest.examples, "examples"),
                    ...getItem(rest.type, "type"),
                    ...getItem(rest.format, "format"),
                    ...getItem(rest.description, 'description')
                };
            }
            if (state.permissions.length && !user)
                return undefined;
            if (state.permissions.length && !user.permissions)
                return undefined;
            // if (state.permissions.length && user.permissions.permissions.includes[])
            return result;
        },
        // provides an output description
        description(description) {
            state.description = description;
            return Give(result, result, state);
        },
        // describe permissions
        for(permission) {
            state.permissions.push(permission);
            state.previous = 'permissions';
            return Give(result, user, state);
        },
        // adds another of the previous thing
        or(item) {
            if (state.previous)
                state[state.previous].push(item);
            return Give(result, user, state);
        },
        // adds another of the previous thing
        and(item) {
            if (state.previous)
                state[state.previous].push(item);
            return Give(result, user, state);
        },
        // sets the model
        as(model) {
            state.type.push(model);
            state.previous = 'type';
            return Give(result, user, state);
        },
        // sets the format
        format(format) {
            state.format = format;
            state.previous = 'format';
            return Give(result, user, state);
        },
        // sets the minimum
        min() {
            return Give(result, user, state);
        },
        // sets the minimum
        max() {
            return Give(result, user, state);
        },
        // sets an example
        example(example) {
            state.examples.push(example);
            state.previous = 'examples';
            return Give(result, user, state);
        },
        colorEnum(dictionary) {
            state.format = Enum;
            state.enumColors = dictionary;
            return Give(result, user, state);
        }
    };
}
/**
 * generates a description of an output function'
 * @param fn output function to describe
 */
export function Describe(fn) {
    try {
        const keys = Object.keys(fn({}, {}));
        const testObj = createTestObject(keys);
        const obj = Object.entries(fn(testObj, {}))
            .reduce((o, [key, value]) => {
            // If key is not a Loy object
            if (value == Test)
                o[key] = {};
            // if key  id loy object
            else
                o[key] = value;
            return o;
        }, {});
        return obj;
    }
    catch (error) {
        return {};
    }
}
export const give = Give;
export const describe = Describe;
