// loy/build/main/index.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("./util");
exports.Test = '<>!-!_LOI_GENERATE_DESCRIPTIONS_!-!<>';
exports.DateItem = 'loi_internal_Date';
exports.Url = 'loi_internal_Url';
exports.File = 'loi_internal_File';
exports.Image = 'loi_internal_Image';
exports.Enum = 'loi_internal_Enum';
/**
 * Return a result
 * @param result expected result
 * @param user user object
 * @param state state
 */
function Give(result, user, state = { permissions: [], examples: [], type: [] }) {
    return {
        // returns value
        ok() {
            if (result === exports.Test) {
                const { previous } = state, rest = __rest(state, ["previous"]);
                return Object.assign(Object.assign(Object.assign(Object.assign(Object.assign({}, util_1.getItem(rest.permissions, "permissions")), util_1.getItem(rest.examples, "examples")), util_1.getItem(rest.type, "type")), util_1.getItem(rest.format, "format")), util_1.getItem(rest.description, 'description'));
            }
            if (state.permissions.length && !user)
                return undefined;
            if (state.permissions.length && !user.permissions)
                return undefined;
            // if (state.permissions.length && user.permissions.permissions.includes[])
            return result;
        },
        // provides an output description
        description(description) {
            state.description = description;
            return Give(result, result, state);
        },
        // describe permissions
        for(permission) {
            state.permissions.push(permission);
            state.previous = 'permissions';
            return Give(result, user, state);
        },
        // adds another of the previous thing
        or(item) {
            if (state.previous)
                state[state.previous].push(item);
            return Give(result, user, state);
        },
        // adds another of the previous thing
        and(item) {
            if (state.previous)
                state[state.previous].push(item);
            return Give(result, user, state);
        },
        // sets the model
        as(model) {
            state.type.push(model);
            state.previous = 'type';
            return Give(result, user, state);
        },
        // sets the format
        format(format) {
            state.format = format;
            state.previous = 'format';
            return Give(result, user, state);
        },
        // sets the minimum
        min() {
            return Give(result, user, state);
        },
        // sets the minimum
        max() {
            return Give(result, user, state);
        },
        // sets an example
        example(example) {
            state.examples.push(example);
            state.previous = 'examples';
            return Give(result, user, state);
        },
        colorEnum(dictionary) {
            state.format = exports.Enum;
            state.enumColors = dictionary;
            return Give(result, user, state);
        }
    };
}
exports.Give = Give;
/**
 * generates a description of an output function'
 * @param fn output function to describe
 */
function Describe(fn) {
    try {
        const keys = Object.keys(fn({}, {}));
        const testObj = util_1.createTestObject(keys);
        const obj = Object.entries(fn(testObj, {}))
            .reduce((o, [key, value]) => {
            // If key is not a Loy object
            if (value == exports.Test)
                o[key] = {};
            // if key  id loy object
            else
                o[key] = value;
            return o;
        }, {});
        return obj;
    }
    catch (error) {
        return {};
    }
}
exports.Describe = Describe;
exports.give = Give;
exports.describe = Describe;
